x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/frequentitemsets/TransactionStore.java

```diff
@@ -0,0 +1,769 @@
+                other.items.get(oldId, scratchBytesRef);
+                long newId = items.add(scratchBytesRef);
+                long oldCount = other.itemCounts.get(oldId);
+                // if item already exists in this.items it is a rewrite
+                if (newId < 0) {
+                    newId = -1 * (newId + 1);
+                } else if (newId >= itemCounts.size()) {
+                    itemCounts = bigArrays.resize(itemCounts, itemCounts.size() + CAPACITY_INCREMENT);
+                }
+
+                // reuse the old counter structure to remember how to rewrite
+                other.itemCounts.set(oldId, newId);
+                itemCounts.increment(newId, oldCount);
+            }
+        }
+
+        // early closing, so memory can be freed before we allocate new memory below
+        other.items.close();
+
+        // merge transactions
+        for (int i = 0; i < other.transactions.capacity(); ++i) {
+            long oldId = other.transactions.id(i);
+            if (oldId >= 0) {
+                other.transactions.get(oldId, scratchBytesRef);
+                scratchBytesRefDataInput.reset(scratchBytesRef.bytes, scratchBytesRef.offset, scratchBytesRef.length);
+                scratchTransactionBytesRefBuilder.clear();
+
+                while (scratchBytesRefDataInput.eof() == false) {
+                    long item = scratchBytesRefDataInput.readVLong();
+
+                    // rewrite and add to builder
+                    writeSignedVLong(scratchTransactionBytesRefBuilder, other.itemCounts.get(item));
+                }
+                long newId = transactions.add(scratchTransactionBytesRefBuilder.get());
+                if (newId < 0) {
+                    newId = -1 * (newId + 1);
+                } else if (newId >= transactionCounts.size()) {
+                    transactionCounts = bigArrays.resize(transactionCounts, transactionCounts.size() + CAPACITY_INCREMENT);
+                }
+                long oldCount = other.transactionCounts.get(oldId);
+                transactionCounts.increment(newId, oldCount);
+            }
+        }
+
+        totalItemCount += other.totalItemCount;
+        totalTransactionCount += other.totalTransactionCount;
+        other.close();
+    }
+
+    /**
+     * Prune transactions and items according to the given min count
+     *
+     * Prune rewrites the internal data structures by getting rid of items which are
+     * below the given minSupport.
+     *
+     * Performance: In addition it re-arranges transactions in order to execute a
+     * contains operation as linear scan, see [ITEM-BOW].
+     *
+     * Currently this is only used after merging all shard stores.
+     *
+     * TODO: explore whether we could prune per shard based on a heuristic
+     *
+     * @param minSupport the minimum support an item must have to be kept
+     */
+    public void prune(double minSupport) {
+
+        long minCount = (long) (minSupport * totalTransactionCount);
```

Comment:
Are we concerned about rounding here? (I genuinely don't know).

It seems like there could be whacky edge cases here.